pubky-homeserver/src/data_directory/config_toml.rs

```
@@ -70,6 +70,12 @@ pub struct GeneralToml {
     pub user_storage_quota_mb: u64,
 }
 
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct LoggingToml {
+    pub level: Option<String>,
+    pub filters: Option<Vec<String>>,
```

- These shouldn't be Options. In this config, both should be mandatory. You can set defaults in `config.default.toml` so devs/users don't need to supply them while testing/running the app.
- Validate the values properly. Not every string is a log level. Not every string is a filter. serde is your friend. See other types.
- What happens if both level and filters are supplied? Do they work in combination?

---

> Filters may set more a verbose level specifically for module.

Do they work too when you set a more restrictive level?

> I've never seen RUST_LOG validating values. And certainly it is not possible for modules.

Use the EnvFilter default parser.

```rust
tracing_subscriber::filter::EnvFilter::builder()
            .parse("pubky_homeserver=debug,tower_http=debug")
```

---

> Validate the values properly. Not every string is a log level. Not every string is a filter. serde is your friend. See other types.

Generally these values are being validated later upon subscriber initialization

```rust
                if let Some(level) = &context.config_toml.logging.level {
                    filter = filter.add_directive(level.parse()?);
                };

                // Add any specific filters
                if let Some(filters) = &context.config_toml.logging.filters {
                    for filter_str in filters {
                        filter = filter.add_directive(filter_str.parse()?);
                    }
                };
```      

Should I write serialization/deserialization functions because tracing:Level and LevelFilter are not supporting traits?

---

> Generally these values are being validated later upon subscriber initialization

You want to push validation as much as possible to the ConfigToml. The logging params are not the best example for this but it helps simplify the code. Failing early is good.

> Should I write serialization/deserialization functions because tracing:Level and LevelFilter are not supporting traits?

Checkout the `data_directory/domain.rs` file. It shows a good example on how to wrap a type (tracing::Level in this case) and validate/serialize/deserialize it.